tests/loggen/loggen.c

```diff
@@ -609,6 +600,49 @@ gint active_finished;
 gint connect_finished;
 guint64 sum_count;
 
+/* Multithreaded OpenSSL */
+GMutex **ssl_lock_cs;
+
+static void
+ssl_threadid_callback(CRYPTO_THREADID *tid)
+{
+  CRYPTO_THREADID_set_pointer(tid, g_thread_self());
+}
+
+static void
+ssl_locking_callback(int mode, int n, const char *file, int line)
+{
+  if (mode & CRYPTO_LOCK)
+    g_mutex_lock(ssl_lock_cs[n]);
+  else
+    g_mutex_unlock(ssl_lock_cs[n]);
+}
+
+static void
+ssl_thread_setup(void)
+{
+  int i;
+  ssl_lock_cs = g_malloc(CRYPTO_num_locks() * sizeof(GMutex *));
+
+  for (i = 0; i < CRYPTO_num_locks(); ++i)
+    ssl_lock_cs[i] = g_mutex_new();
+
+  CRYPTO_THREADID_set_callback(ssl_threadid_callback);
+  CRYPTO_set_locking_callback(ssl_locking_callback);
+}
+
+static void
+ssl_thread_cleanup(void)
+{
+  int i;
+  CRYPTO_set_locking_callback(NULL);
+
+  for (i = 0; i < CRYPTO_num_locks(); ++i)
+    g_mutex_free(ssl_lock_cs[i]);
+
+  g_free(ssl_lock_cs);
+}
+
 gpointer
```

Comment:
Until now, loggen was completely independent from the core, and only
depends on the std library, to be able to run on boxes that do not run
syslog-ng.

This could be reconsidered though.
On Aug 31, 2016 2:10 PM, "László Várady" <notifications@github.com> wrote:

> In tests/loggen/loggen.c
> <https://github.com/balabit/syslog-ng/pull/1186#discussion_r76975694>:
>
> > +  CRYPTO_THREADID_set_callback(ssl_threadid_callback);
> > +  CRYPTO_set_locking_callback(ssl_locking_callback);
> > +}
> > +
> > +static void
> > +ssl_thread_cleanup(void)
> > +{
> > +  int i;
> > +  CRYPTO_set_locking_callback(NULL);
> > +
> > +  for (i = 0; i < CRYPTO_num_locks(); ++i)
> > +    g_mutex_free(ssl_lock_cs[i]);
> > +
> > +  g_free(ssl_lock_cs);
> > +}
> > +
> >  gpointer
>
> Thanks @bazsi <https://github.com/bazsi>, you are right. I just noticed
> that the same lock functionality exists in lib/crypto.c.
> Maybe I could link against lib/crypto statically and use
> crypto_init_threading() instead. What do you think?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/balabit/syslog-ng/pull/1186/files/112cf4059040d39b922d0fe1ac4c4ab533506d6e#r76975694>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AArldgoFf5sHdVF5zHVTHcuMGc8nyz7dks5qlW8vgaJpZM4JvlAP>
> .
>